### DISCUSSION TITLE
Does Mojo have syntactic macros?

### DISCUSSION THREAD

**Discussion by chwmilcon** [modular/modular#3406] [Q&A]
Does Mojo have syntactic macros? Not parameterization but syntactic macros like Rust. There have been several articles on Medium that show supposed mojo code that is generating code via macros, but they don't seem to compile using version 24 of mojo that I have. The macro's in the articles look like:
```
macro createGetterSetter(name: String, type: String) -> String {
    return """
    func get\(name.capitalized)() -> \(type) {
        return self.\(name)
    }

    func set\(name.capitalized)(value: \(type)) {
        self.\(name) = value
    }
    """
}
```

**Comment by ivellapillil** [comment]
Mojo doesn't have macros. However there are plans to have decorators similar to Python. Mojo also allows code to be executed at compile time, for example when you use `alias` assignment. 

At the moment custom decorators are not supported,  but it is expected that it would cover some of the needs of macros.

**Comment by Moosems** [comment]
I really hope the team doesn't add macros. Just make a function instead. The code you're seeing is probably AI generated as macros have never existed in Mojo to my knowledge

**Reply by chwmilcon** [reply]
You are probably correct that they are AI generated articles, I asked because the two articles seemed well written and gave different examples, but both using the same, apparently non-existent syntax.  The code that I gave as an example could be done as simple functions, but syntactic macros are far more powerful than simply replacing a function. Looking at what you can do with them in languages like Lisp or Rust shows that.

**Comment by Brian-M-J** [comment]
If Python adds macros, as suggested by [PEP 638](https://peps.python.org/pep-0638/), then it's guaranteed that Mojo will also add them. Till then, I don't know what the team's plans are for them.

There seems to be [support](https://github.com/modularml/mojo/discussions/244#discussioncomment-6008071) for user-defined statements, though.

**Reply by Moosems** [reply]
Despite the goal being to eliminate confusion I can only see this making the language more difficult to use and dividing the user base significantly
